src/redisraft.c

```diff
@@ -1008,6 +1008,105 @@ static int cmdRaftRandom(RedisModuleCtx *ctx,
     return REDISMODULE_OK;
 }
 
+#ifdef HAVE_TLS
+/* Callback for passing a keyfile password stored as a char * to OpenSSL,  copied from redis */
+static int tlsPasswordCallback(char *buf, int size, int rwflag, void *u)
+{
+    const char *pass = u;
+    size_t pass_len;
+
+    if (!pass) {
+        return -1;
+    }
+    pass_len = strlen(pass);
+    if (pass_len > (size_t) size) {
+        return -1;
+    }
+    memcpy(buf, pass, pass_len);
+
+    return (int) pass_len;
+}
+
+SSL_CTX *generateSSLContext(RedisModuleCtx *ctx, RedisRaftCtx *rr)
+{
+    SSL_CTX *ssl_ctx = SSL_CTX_new(SSLv23_client_method());
+    if (!ssl_ctx) {
+        LOG_ERROR("REDIS_SSL_CTX_CREATE_FAILED");
+        return NULL;
+    }
+
+    SSL_CTX_set_options(ssl_ctx, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3);
+    SSL_CTX_set_verify(ssl_ctx, SSL_VERIFY_PEER, NULL);
+
+    if ((rr->config->tls_cert != NULL && rr->config->tls_key == NULL) ||
+            (rr->config->tls_key != NULL && rr->config->tls_cert == NULL)) {
+        LOG_ERROR("REDIS_SSL_CTX_CERT_KEY_REQUIRED");
+        goto error;
+    }
+
+    if (!SSL_CTX_load_verify_locations(ssl_ctx, rr->config->tls_ca_cert, NULL)) {
+        unsigned long e = ERR_peek_last_error();
+        LOG_ERROR("SSL_CTX_load_verify_locations(): %s", ERR_reason_error_string(e));
+        goto error;
+    }
+
+    if (!SSL_CTX_use_certificate_chain_file(ssl_ctx, rr->config->tls_cert)) {
+        unsigned long e = ERR_peek_last_error();
+        LOG_ERROR("SSL_CTX_use_certificate_chain_file(): tls_cert = %s, err = %s", rr->config->tls_cert, ERR_reason_error_string(e));
+        goto error;
+    }
+    if (!SSL_CTX_use_PrivateKey_file(ssl_ctx, rr->config->tls_key, SSL_FILETYPE_PEM)) {
+        unsigned long e = ERR_peek_last_error();
+        LOG_ERROR("SSL_CTX_use_PrivateKey_file(): tls_key = %s, err = %s", rr->config->tls_key, ERR_reason_error_string(e));
+        goto error;
+    }
+
+    if (rr->config->tls_key_pass && *rr->config->tls_key_pass != '\0') {
+        SSL_CTX_set_default_passwd_cb(ssl_ctx, tlsPasswordCallback);
+        SSL_CTX_set_default_passwd_cb_userdata(ssl_ctx, (void *) rr->config->tls_key_pass);
+    }
+
+    return ssl_ctx;
+
+error:
+    SSL_CTX_free(ssl_ctx);
+    return NULL;
+}
+#endif
+
+
+void handleConfigChangeEvent(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64_t subevent, void *data)
+{
+    LOG_ERROR("handleConfigChangeEvent: enter");
+    if (eid.id != REDISMODULE_EVENT_CONFIG || subevent != REDISMODULE_SUBEVENT_CONFIG_CHANGE) {
+        return;
+    }
+
+#ifdef HAVE_TLS
+    if (!redis_raft.config->tls_enabled || redis_raft.config->tls_manual) {
+        return;
+    }
+
+    RedisModuleConfigChangeV1 *ei = data;
+
+    for (unsigned int i = 0; i < ei->num_changes; i++) {
+        if (strcmp("tls-ca-cert-file", ei->config_names[i]) == 0 ||
+                strcmp("tls-key-file", ei->config_names[i]) == 0 ||
+                strcmp("tls-cert-file", ei->config_names[i]) == 0) {
```

Comment:
I think we'll need to detect `tls-client-*` family changes as well.

Reply:
correct.

Reply:
adde